### PR TITLE
src: Load plugins after netlink PM is available

### PR DESCRIPTION
**Pull request by ossama-othman**
Defer plugin load until after the generic netlink family for the
netlink path manager is ready for use since the recent plugin
framework updates make it possible for path management related calls
to be invoked during plugin initialization.  This also effectively
ensures that the mptcpd path manager bridge is always ready, obviating
the need to poll for readiness by plugins.

### CODE REVIEW

**Comment by ossama-othman** [comment]
Not quite ready for review.

File: src/path_manager.c [hunk 1]
```diff
@@ -1154,17 +1194,21 @@ struct mptcpd_pm *mptcpd_pm_create(struct mptcpd_config const *config)
...
-                                    NULL) == 0
+                                    data,
+                                    l_free) == 0
             || !l_genl_request_family(pm->genl,
                                       name,
                                       family_appeared,
-                                      pm,
-                                      NULL)) {
+                                      data,
+                                      l_free)) {
```
Comment (mjmartineau) [inline src/path_manager.c:1211]: ..._and_ here looks a little funny (in terms of how the ELL APIs work), but as far as I can see `l_genl_add_family_watch()` will not free `data` if it returns 0, and `l_genl_request_family()` will not free `data` if it returns `false`. But that also implies that `l_free(data)` is needed in the error handling block of code below.

Comment (ossama-othman) [reply src/path_manager.c:1211]: Good catch!  I'll add an `l_free()` to the error handling block.

Comment (ossama-othman) [reply src/path_manager.c:1211]: Looking at the code again, I am concerned about a potential double-`free()` on `data`.  I'll take a closer look.

Comment (ossama-othman) [reply src/path_manager.c]: Commit 6b9de58 simplifies this patch set and addresses the memory management issues.  There is a longer a need to dynamically allocate and manage a convenience structure like  `family_cb_data` that introduced earlier.

**Review by mjmartineau — APPROVED** [review]

File: src/path_manager.c [hunk 2]
```diff
@@ -988,6 +986,22 @@ static void complete_pm_init(struct mptcpd_pm *pm)
...
+         *      @c mptcpd_plugin_unload() is called.
+         */
+        if (!mptcpd_plugin_load(pm->config->plugin_dir,
+                                pm->config->default_plugin,
+                                pm)) {
+                l_error("Unable to load path manager plugins.");
+
+                mptcpd_pm_destroy(pm);
+
+                exit(EXIT_FAILURE);
```
Comment (ossama-othman) [inline src/path_manager.c:1003]: I ran a quick leak check through `valgrind`, and found a small memory leak at process exit when this condition (plugin load failure) is triggered:
```
==11756== 40 bytes in 1 blocks are definitely lost in loss record 19 of 25
==11756==    at 0x4839809: malloc (vg_replace_malloc.c:307)
==11756==    by 0x48608FD: l_malloc (util.c:62)
==11756==    by 0x4865D26: watch_add (main.c:144)
==11756==    by 0x4867158: l_io_new (io.c:191)
==11756==    by 0x486BC54: l_genl_new (genl.c:1140)
==11756==    by 0x10D59F: mptcpd_pm_create (path_manager.c:1166)
==11756==    by 0x10B62E: main (mptcpd.c:58)
```
I'll create a separate pull request for the fix.

**Pull request merged**